org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/format/LSPFormatter.java

```diff
@@ -54,17 +57,17 @@ public CompletableFuture<Optional<VersionedEdits>> requestFormatting(@NonNull ID
 		// TODO: Could refine this algorithm: at present this grabs the first non-null response but the most functional
 		// implementation (if a text selection is present) would try all the servers in turn to see if they supported
 		// range formatting, falling back to a full format if unavailable
+		var builder = new VersionedBuilder<List<? extends TextEdit>, @NonNull VersionedEdits>(document, VersionedEdits::new);
```

Comment:
What a builder and not directly a VersionedEdit ?

Reply:
I think it fits nicely our problem:

- We need to record the document state right before sending the request
- That recorded value should be added to the returned data to create a Versioned<T> than then you can use

Having the builder clearly makes this visible in the code. We could of course create already a Versioned<T> that we then later complete, and then gets usable, but it does not enforce that the pattern is followed.

Reply:
> We need to record the document state right before sending the request

That can be achieved by directly instantiating a VersionedEdits here (Versioned should probably contain the document as field by the way)

> That recorded value should be added to the returned data to create a Versioned than then you can use

Yes, but something like `versionedEdits.apply` could do it (`apply` being `public void apply(List<TextEdit>>)` which would check the doc version before performing the actual change.

So code would look like
```java
VersionedEdits edits = new VersionedEdits(document);
return executor.computeFirst((w, ls) -> {
		final ServerCapabilities capabilities = w.getServerCapabilities();
		if (isDocumentRangeFormattingSupported(capabilities)
				&& !(isDocumentFormattingSupported(capabilities)
						&& textSelection.getLength() == 0)) {
			return ls.getTextDocumentService().rangeFormatting(rangeParams).thenApply(builder::build);
		}
		return ls.getTextDocumentService().formatting(params).thenApply(edits::apply);
});
```
This has the advantage of avoid intermediary object, IMO making the code clearer.

Reply:
> But what if you would use computeAll?

Indeed, this case would require multiple VersionedEdits I guess. However, do we already have this case yet? I'd like to avoid introducing one solution before the problem exists ;)

> `VersionedEdits edits = new VersionedEdits(document);` would also be an extra object as in the builder.

Yes, but I think most people are more comfortable with `new` over `Builder` unless it's proven a Builder makes things simpler. 
```java
VersionedEdits edits = new VersionedEdits(document);
edits.apply(textEdits);
// or
edits.addEdits(textEdits);
edits.apply();
```
seems to cover the current need (assuming we don't need it with computeAll at the moment, which may be a wrong assumption), with just 1 method addition that is relatively straightforward to guess with completion.

Reply:
Agree, if the builder does not solve a problem, and most people are indeed more confortable with `new` over `Builder`, that is fine for me, it was indeed my proposal in the first commit https://github.com/eclipse/lsp4e/pull/542/commits/2f4924315972e64db99e8df4be623ca8e24fd255
Just 
`		long modificationStamp = DocumentUtil.getDocumentModificationStamp(document);`
to record the timestamp and then
`edits -> new VersionedEdits(modificationStamp, edits, document)`
inside the lamba. No new pattern and it is straghforward. You need to pass three arguments instead of one.

Shall we go for that one?

Note that as I tried to explain above, even if computeFirst returns only the first result, it does not guarantee that only one lambda is executed, so
```
VersionedEdits edits = new VersionedEdits(document);
return executor.computeFirst((w, ls) -> {
                ...
		return ls.getTextDocumentService().formatting(params).thenApply(edits::apply);
});
```
needs `edits::apply` to be thread safe, and even then the versioned edit could contain edits from more than one server.